core/src/connection.rs

```diff
@@ -19,27 +19,37 @@
 // DEALINGS IN THE SOFTWARE.
 
 use crate::multiaddr::{Multiaddr, Protocol};
+use std::fmt;
+use std::fmt::Debug;
+use std::sync::atomic::{AtomicU64, Ordering};
+
+static NEXT_CONNECTION_ID: AtomicU64 = AtomicU64::new(0);
 
 /// Connection identifier.
-#[derive(Debug, Copy, Clone, Hash, PartialEq, Eq, PartialOrd, Ord)]
-pub struct ConnectionId(usize);
+#[derive(Debug, Copy, Clone, Hash, PartialEq, Eq)]
+pub struct ConnectionId(u64);
 
 impl ConnectionId {
     /// Creates a `ConnectionId` from a non-negative integer.
     ///
     /// This is primarily useful for creating connection IDs
     /// in test environments. There is in general no guarantee
     /// that all connection IDs are based on non-negative integers.
+    #[deprecated(note = "IDs must be unique and should not be constructed directly.")]
     pub fn new(id: usize) -> Self {
-        Self(id)
+        Self(id as u64)
     }
 }
 
-impl std::ops::Add<usize> for ConnectionId {
-    type Output = Self;
+impl Default for ConnectionId {
```

Comment:
Ignoring the fact that there has already been a `ConnectionId::new` and that that method takes a parameter.

My intuition for `Default` would be that two `Default::default()` would equal. Is that idiomatic Rust or only my intuition.

This is not particularly important. Fine with keeping it as is.

Reply:
Good question. I think I agree with you, now that you are pointing it out!

We can easily replace it with a public `next_unused` constructor.